File: backend/src/models/YearlyRecord.ts
import { db } from '../global/firebase';
import { Timestamp, FieldValue } from 'firebase-admin/firestore';

export interface YearlyRecordInput {
  assetType: string;
  year: number;
  prevYearTotal: number;
  amountInvested: number;
  stockValue: number;
  cashBalance: number;
  foreignValueTwd: number;
  returnAmount: number;
  returnRate: number;
  settledAt: string; // ISO string
  note?: string;
}

export class YearlyRecord {
  id!: string; // {assetType}_{year}
  assetType!: string;
  year!: number;
  prevYearTotal!: number;
  amountInvested!: number;
  stockValue!: number;
  cashBalance!: number;
  foreignValueTwd!: number;
  returnAmount!: number;
  returnRate!: number;
  settledAt!: string;
  note!: string;
  createdAt!: string;

  private static readonly col = db.collection('yearly_records');

  static async findAll(assetType = 'tw_stock'): Promise<YearlyRecord[]> {
    const snap = await this.col
      .where('asset_type', '==', assetType)
      .orderBy('year', 'asc')
      .get();
    return snap.docs.map(doc => YearlyRecord.fromSnapshot(doc));
  }

  static async findByYear(assetType: string, year: number): Promise<YearlyRecord | null> {
    const doc = await this.col.doc(`${assetType}_${year}`).get();
    return doc.exists ? YearlyRecord.fromSnapshot(doc) : null;
  }

  static async create(input: YearlyRecordInput): Promise<YearlyRecord> {
    const docId = `${input.assetType}_${input.year}`;
    const ref = this.col.doc(docId);
    await ref.set({
      asset_type:       input.assetType,
      year:             input.year,
      prev_year_total:  input.prevYearTotal,
      amount_invested:  input.amountInvested,
      stock_value:      input.stockValue,
      cash_balance:     input.cashBalance,
      foreign_value_twd: input.foreignValueTwd,
      return_amount:    input.returnAmount,
      return_rate:      input.returnRate,
      settled_at:       Timestamp.fromDate(new Date(input.settledAt)),
      note:             input.note ?? '',
      created_at:       FieldValue.serverTimestamp(),
    });
    const doc = await ref.get();
    return YearlyRecord.fromSnapshot(doc);
  }

  static async update(assetType: string, year: number, input: Partial<YearlyRecordInput>): Promise<YearlyRecord | null> {
    const ref = this.col.doc(`${assetType}_${year}`);
    const doc = await ref.get();
    if (!doc.exists) return null;

    const patch: Record<string, unknown> = {};
    if (input.prevYearTotal   !== undefined) patch['prev_year_total']   = input.prevYearTotal;
    if (input.amountInvested  !== undefined) patch['amount_invested']   = input.amountInvested;
    if (input.stockValue      !== undefined) patch['stock_value']       = input.stockValue;
    if (input.cashBalance     !== undefined) patch['cash_balance']      = input.cashBalance;
    if (input.foreignValueTwd !== undefined) patch['foreign_value_twd'] = input.foreignValueTwd;
    if (input.returnAmount    !== undefined) patch['return_amount']     = input.returnAmount;
    if (input.returnRate      !== undefined) patch['return_rate']       = input.returnRate;
    if (input.settledAt       !== undefined) patch['settled_at']        = Timestamp.fromDate(new Date(input.settledAt));
    if (input.note            !== undefined) patch['note']              = input.note;

    await ref.update(patch);
    const updated = await ref.get();
    return YearlyRecord.fromSnapshot(updated);
  }

  private static fromSnapshot(doc: FirebaseFirestore.DocumentSnapshot): YearlyRecord {
    const d = doc.data()!;
    const r = new YearlyRecord();
    r.id              = doc.id;
    r.assetType       = d['asset_type'];
    r.year            = d['year'];
    r.prevYearTotal   = d['prev_year_total'];
    r.amountInvested  = d['amount_invested'];
    r.stockValue      = d['stock_value'];
    r.cashBalance     = d['cash_balance'];
    r.foreignValueTwd = d['foreign_value_twd'];
    r.returnAmount    = d['return_amount'];
    r.returnRate      = d['return_rate'];
    const sa          = d['settled_at'];
    r.settledAt       = sa instanceof Timestamp ? sa.toDate().toISOString() : '';
    r.note            = d['note'] ?? '';
    const ca          = d['created_at'];
    r.createdAt       = ca instanceof Timestamp ? ca.toDate().toISOString() : new Date().toISOString();
    return r;
  }
}

File: Back-End/shioaji-service/models/yearly_record.py
from __future__ import annotations
import asyncio
from typing import Optional
from google.cloud.firestore import SERVER_TIMESTAMP
from lib.firebase import get_db

COL = "yearly_records"


def _ts_to_iso(ts) -> Optional[str]:
    return ts.isoformat() if hasattr(ts, "isoformat") else None


def _from_doc(doc) -> dict:
    d = doc.to_dict()
    return {
        "id": doc.id,
        "assetType": d.get("asset_type", ""),
        "year": d.get("year", 0),
        "prevYearTotal": d.get("prev_year_total", 0),
        "amountInvested": d.get("amount_invested", 0),
        "stockValue": d.get("stock_value", 0),
        "cashBalance": d.get("cash_balance", 0),
        "foreignValueTwd": d.get("foreign_value_twd", 0),
        "returnAmount": d.get("return_amount", 0),
        "returnRate": d.get("return_rate", 0),
        "settledAt": _ts_to_iso(d.get("settled_at")),
        "note": d.get("note", ""),
        "createdAt": _ts_to_iso(d.get("created_at")),
    }


async def find_all(asset_type: str = "tw_stock") -> list[dict]:
    def _run():
        snap = (
            get_db().collection(COL)
            .where("asset_type", "==", asset_type)
            .order_by("year")
            .get()
        )
        return [_from_doc(d) for d in snap]
    return await asyncio.to_thread(_run)


async def find_by_year(asset_type: str, year: int) -> Optional[dict]:
    def _run():
        doc = get_db().collection(COL).document(f"{asset_type}_{year}").get()
        return _from_doc(doc) if doc.exists else None
    return await asyncio.to_thread(_run)


async def create(data: dict) -> dict:
    def _run():
        from datetime import datetime, timezone
        asset_type = data["assetType"]
        year = data["year"]
        ref = get_db().collection(COL).document(f"{asset_type}_{year}")
        settled_ts = None
        if data.get("settledAt"):
            settled_ts = datetime.fromisoformat(data["settledAt"].replace("Z", "+00:00"))
        ref.set({
            "asset_type": asset_type,
            "year": year,
            "prev_year_total": data.get("prevYearTotal", 0),
            "amount_invested": data.get("amountInvested", 0),
            "stock_value": data.get("stockValue", 0),
            "cash_balance": data.get("cashBalance", 0),
            "foreign_value_twd": data.get("foreignValueTwd", 0),
            "return_amount": data.get("returnAmount", 0),
            "return_rate": data.get("returnRate", 0),
            "settled_at": settled_ts,
            "note": data.get("note", ""),
            "created_at": SERVER_TIMESTAMP,
        })
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)


async def update(asset_type: str, year: int, patch: dict) -> Optional[dict]:
    def _run():
        from datetime import datetime, timezone
        ref = get_db().collection(COL).document(f"{asset_type}_{year}")
        if not ref.get().exists:
            return None
        data: dict = {}
        if "prevYearTotal"   in patch: data["prev_year_total"]   = patch["prevYearTotal"]
        if "amountInvested"  in patch: data["amount_invested"]   = patch["amountInvested"]
        if "stockValue"      in patch: data["stock_value"]       = patch["stockValue"]
        if "cashBalance"     in patch: data["cash_balance"]      = patch["cashBalance"]
        if "foreignValueTwd" in patch: data["foreign_value_twd"] = patch["foreignValueTwd"]
        if "returnAmount"    in patch: data["return_amount"]     = patch["returnAmount"]
        if "returnRate"      in patch: data["return_rate"]       = patch["returnRate"]
        if "note"            in patch: data["note"]              = patch["note"]
        if "settledAt"       in patch:
            data["settled_at"] = datetime.fromisoformat(patch["settledAt"].replace("Z", "+00:00"))
        if data:
            ref.update(data)
        return _from_doc(ref.get())
    return await asyncio.to_thread(_run)
